src/main/java/com/sucy/skill/listener/ProjectileListener.java

```diff
@@ -0,0 +1,62 @@
+package com.sucy.skill.listener;
+
+import com.sucy.skill.api.event.ProjectileTickEvent;
+import com.sucy.skill.task.ProjectileTickTask;
+import com.sucy.skill.thread.MainThread;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Projectile;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.entity.ProjectileHitEvent;
+import org.bukkit.event.entity.ProjectileLaunchEvent;
+import org.bukkit.projectiles.ProjectileSource;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Listener to throw custom {@link ProjectileTickEvent}
+ */
+public class ProjectileListener extends SkillAPIListener {
+    private static final List<UUID> flyingProjectiles = new ArrayList<>();
+
+    @Override
+    public void cleanup() {
+        ProjectileListener.flyingProjectiles.clear();
+    }
+
+    /**
+     * Marks projectiles as "flying"
+     */
+    @EventHandler
+    public void onLaunch(ProjectileLaunchEvent event) {
+        ProjectileSource shooter = event.getEntity().getShooter();
+        if (!(shooter instanceof LivingEntity)) return;
+
+        flyingProjectiles.add(event.getEntity().getUniqueId());
+        MainThread.register(new ProjectileTickTask((LivingEntity) event.getEntity().getShooter(), event.getEntity()));
```

Comment:
I'm a little weary of always scheduling the tick task here regardless of if it will be applicable at all. I kind of feel like if we were able to find a way to only run this task if there is a valid tick trigger on the firing player, that would be better. I'm not entirely sure the best implementation for this kind of thing, I'm just a little concerned about the potential performance hit that could be caused by tracking potentially dozens-hundreds of projectiles every tick like this.

Reply:
Yeah that's a good point. I'm not entirely sure the best solution at the moment, but it's worth giving a little more thought. At the end of the day, we may just go with exactly what you have and hope it doesn't cause too many issues with performance.